base/arraymath.jl

```diff
@@ -470,14 +476,18 @@ for (f, f!, fp, op) = ((:cumsum, :cumsum!, :cumsum_pairwise!, :+),
     end
 
     @eval function ($f!)(result::AbstractVector, v::AbstractVector)
-        n = length(v)
+        li = linearindices(v)
+        li != linearindices(result) && throw(BoundsError())
```

Comment:
Should this be an `DimensionMismatch` and include a helpful message about the nature of the problem?